### PR TITLE
Normalize lazilly for sized

### LINKED ISSUE #31299: Bogus error in beta and nightly: recursive type has infinite size
```rust
trait Front {
    type Back;
}

impl<T> Front for Vec<T> {
    type Back = Vec<T>;
}

struct PtrBack<T: Front>(*mut T::Back);

struct M(PtrBack<Vec<M>>);
 
fn main() {
    println!("{}", std::mem::size_of::<M>());
}
```

This program prints `8` in Stable (1.5.0), but in Beta and Nightly compilation fails with:

```
<anon>:11:1: 11:27 error: recursive type `M` has infinite size [E0072]
<anon>:11 struct M(PtrBack<Vec<M>>);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~
<anon>:11:1: 11:27 help: see the detailed explanation for E0072
<anon>:11:1: 11:27 help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `M` representable
<anon>:11:1: 11:27 note: type `M` is embedded within `PtrBack<collections::vec::Vec<M>>`...
<anon>:11:1: 11:27 note: ...which in turn is embedded within `PtrBack<collections::vec::Vec<M>>`...
<anon>:11:1: 11:27 note: ...which in turn is embedded within `M`, completing the cycle.
```

From 8 bytes to infinity would seem to be a serious memory usage regression.

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a bit of a hack, but if we eagerly expand structural type obligations, we can resolve #31299. 

r? @arielb1

### CODE REVIEW

**Comment by arielb1** [comment]
This does not handle newtyped pointers (`Unique<T>` etc.), right? I am not sure we want to land something like this.

File: src/librustc/traits/select.rs [hunk 1]
```diff
@@ -1783,8 +1793,8 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             ty::TyArray(element_ty, _) => {
                 // [T; n]
                 match bound {
-                    ty::BoundCopy => ok_if(vec![element_ty]),
-                    ty::BoundSized => ok_if(Vec::new()),
+                    ty::BoundCopy => self.builtin_bound_ok_if(bound, vec![element_ty]),
+                    ty::BoundSized => self.builtin_bound_ok_if(bound, vec![]),
```
Comment (arielb1) [inline src/librustc/traits/select.rs]: This looks a bit heavy.

**Comment by arielb1** [comment]
How does last-field-only not break our self-recursive struct checking?

**Comment by nikomatsakis** [comment]
@arielb1 

> How does last-field-only not break our self-recursive struct checking?

We have separate code that checks for this. It so happens that if we get a recursive `Sized` error, we will try to detect this case and print it nicely, but it's not the only place that checks for this condition iirc.

When I run that test I get:

```rust
/home/nmatsakis/tmp/x.rs:1:1: 4:2 error: recursive type `Recursive` has infinite size [E0072]
/home/nmatsakis/tmp/x.rs:1 struct Recursive {
/home/nmatsakis/tmp/x.rs:2     this: Recursive,
/home/nmatsakis/tmp/x.rs:3     last: u32
/home/nmatsakis/tmp/x.rs:4 }
/home/nmatsakis/tmp/x.rs:1:1: 4:2 help: run `rustc --explain E0072` to see a detailed explanation
/home/nmatsakis/tmp/x.rs:1:1: 4:2 help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `Recursive` representable
error: aborting due to previous error
```

I can add a test case if you like.

**Comment by nikomatsakis** [comment]
@arielb1 

> This does not handle newtyped pointers (Unique<T> etc.), right? I am not sure we want to land something like this.

No, it doesn't, no structs. Why not land it? Just because it's hacky?

**Comment by nikomatsakis** [comment]
Withdrawing this PR in favor of https://github.com/rust-lang/rust/pull/33138/, which is way better.

**Pull request closed without merging**